paimon-core/src/main/java/org/apache/paimon/tag/TagTimeExpire.java

```diff
@@ -52,24 +59,41 @@ private TagTimeExpire(
         this.callbacks = callbacks;
     }
 
-    public void run() {
+    public List<String> expire() {
         List<Pair<Tag, String>> tags = tagManager.tagObjects();
+        FileIO fileIO = snapshotManager.fileIO();
+        List<String> expired = new ArrayList<>();
         for (Pair<Tag, String> pair : tags) {
             Tag tag = pair.getLeft();
             String tagName = pair.getRight();
             LocalDateTime createTime = tag.getTagCreateTime();
-            Duration timeRetained = tag.getTagTimeRetained();
-            if (createTime == null || timeRetained == null) {
-                continue;
+            if (createTime == null && olderThanTime != null) {
+                FileStatus tagFileStatus;
+                try {
+                    tagFileStatus = fileIO.getFileStatus(tagManager.tagPath(tagName));
+                } catch (IOException e) {
+                    LOG.warn("Tag path {} not exist, skip expire it.", tagManager.tagPath(tagName));
+                    continue;
+                }
+                createTime = DateTimeUtils.toLocalDateTime(tagFileStatus.getModificationTime());
             }
-            if (LocalDateTime.now().isAfter(createTime.plus(timeRetained))) {
+            Duration timeRetained = tag.getTagTimeRetained();
+            if ((timeRetained != null && LocalDateTime.now().isAfter(createTime.plus(timeRetained)))
+                    || (olderThanTime != null && olderThanTime.isAfter(createTime))) {
                 LOG.info(
                         "Delete tag {}, because its existence time has reached its timeRetained of {}.",
```

Comment:
LOG.info(
                        "Delete tag {}, because its existence time has reached its timeRetained of {} or" +
                                " its createTime {} is olderThan olderThanTime {}.",
                        tagName,
                        timeRetained,
                        createTime,
                        olderThanTime);